benches/dapol.rs

```diff
@@ -1,175 +1,385 @@
-use core::fmt::Debug;
-use criterion::{criterion_group, criterion_main, BatchSize, BenchmarkId, Criterion, SamplingMode};
-use dapol::{
-    utils::get_secret, Dapol, DapolNode, RangeProofPadding, RangeProofSplitting, RangeProvable,
-    RangeVerifiable,
-};
-use digest::Digest;
-use rand::{distributions::Uniform, thread_rng, Rng};
-use smtree::{
-    index::TreeIndex,
-    traits::{ProofExtractable, Rand, Serializable, TypeName},
-};
-use std::time::Duration;
-
-// CONSTANTS
-// ================================================================================================
+use std::path::PathBuf;
+
+use criterion::{criterion_group, criterion_main, SamplingMode};
+use criterion::{BenchmarkId, Criterion};
+use iai_callgrind::{black_box, library_benchmark, library_benchmark_group, main};
+
+use dapol::accumulators::NdmSmt;
+use dapol::{Height, MaxThreadCount};
+
+mod setup;
 
-const TREE_HEIGHTS: [usize; 3] = [16, 24, 32];
-const NUM_LEAVES: [usize; 3] = [1024, 2048, 4096];
+use setup::{NUM_USERS, TREE_HEIGHTS};
 
-// BENCHMARKS
+// BENCHMARKS: CRITERION
 // ================================================================================================
 
-fn build_dapol(c: &mut Criterion) {
+fn bench_build_tree(c: &mut Criterion) -> Result<(), &str> {
```

Comment:
What does Criterion require in terms of return value? Like what is the difference in returning a monad and just `group.finish()`?

Reply:
Returning a `Result` is just for debugging purposes, as there are custom error messages where appropriate. E.g., in the serialize_tree() function, which previously panicked

Reply:
So will Criterion log the error message if you return a `Result` and it wraps an `Err`?

Reply:
Nope, I checked. I'll just leave them out